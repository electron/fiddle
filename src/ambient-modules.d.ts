declare module 'namor';
declare module 'shell-path';
