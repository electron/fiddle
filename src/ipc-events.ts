export enum IpcEvents {
  OPEN_SETTINGS = 'OPEN_SETTINGS',
  LOAD_GIST_REQUEST = 'LOAD_GIST_REQUEST',
  LOAD_ELECTRON_EXAMPLE_REQUEST = 'LOAD_ELECTRON_EXAMPLE_REQUEST',
  FIDDLE_RUN = 'FIDDLE_RUN',
  FIDDLE_PACKAGE = 'FIDDLE_PACKAGE',
  FIDDLE_MAKE = 'FIDDLE_MAKE',
  MONACO_EXECUTE_COMMAND = 'MONACO_EXECUTE_COMMAND',
  MONACO_TOGGLE_OPTION = 'MONACO_TOGGLE_OPTION',
  FS_NEW_FIDDLE = 'FS_NEW_FIDDLE',
  FS_NEW_TEST = 'FS_NEW_TEST',
  FS_OPEN_FIDDLE = 'FS_OPEN_FIDDLE',
  FS_OPEN_TEMPLATE = 'FS_OPEN_TEMPLATE',
  FS_SAVE_FIDDLE = 'FS_SAVE_FIDDLE',
  FS_SAVE_FIDDLE_AS = 'FS_SAVE_FIDDLE_AS',
  FS_SAVE_FIDDLE_GIST = 'FS_SAVE_FIDDLE_GIST',
  FS_SAVE_FIDDLE_FORGE = 'FS_SAVE_FIDDLE_FORGE',
  FS_SAVE_FIDDLE_DIALOG = 'FS_SAVE_FIDDLE_DIALOG',
  FS_SAVE_FIDDLE_ERROR = 'FS_SAVE_FIDDLE_ERROR',
  SHOW_WARNING_DIALOG = 'SHOW_WARNING_DIALOG',
  SHOW_CONFIRMATION_DIALOG = 'SHOW_CONFIRMATION_DIALOG',
  SHOW_WELCOME_TOUR = 'SHOW_WELCOME_TOUR',
  CLEAR_CONSOLE = 'CLEAR_CONSOLE',
  LOAD_LOCAL_VERSION_FOLDER = 'LOAD_LOCAL_VERSION_FOLDER',
  SHOW_LOCAL_VERSION_FOLDER_DIALOG = 'SHOW_LOCAL_VERSION_FOLDER_DIALOG',
  BISECT_COMMANDS_TOGGLE = 'BISECT_COMMANDS_TOGGLE',
  SHOW_INACTIVE = 'SHOW_INACTIVE',
  BEFORE_QUIT = 'BEFORE_QUIT',
  CONFIRM_QUIT = 'CONFIRM_QUIT',
  GET_APP_PATHS = 'GET_APP_PATHS',
  SELECT_ALL_IN_EDITOR = 'SELECT_ALL_IN_EDITOR',
  BLOCK_ACCELERATORS = 'BLOCK_ACCELERATORS',
  SET_SHOW_ME_TEMPLATE = 'SET_SHOW_ME_TEMPLATE',
  CLICK_TITLEBAR_MAC = 'CLICK_TITLEBAR_MAC',
  TASK_BISECT = 'TASK_BISECT',
  TASK_TEST = 'TASK_TEST',
  TASK_DONE = 'TASK_DONE',
  OUTPUT_ENTRY = 'OUTPUT_ENTRY',
  RELOAD_WINDOW = 'RELOAD_WINDOW',
}

export const ipcMainEvents = [
  IpcEvents.FS_SAVE_FIDDLE_DIALOG,
  IpcEvents.FS_SAVE_FIDDLE,
  IpcEvents.SHOW_WARNING_DIALOG,
  IpcEvents.SHOW_CONFIRMATION_DIALOG,
  IpcEvents.SHOW_LOCAL_VERSION_FOLDER_DIALOG,
  IpcEvents.SHOW_INACTIVE,
  IpcEvents.CONFIRM_QUIT,
  IpcEvents.SET_SHOW_ME_TEMPLATE,
  IpcEvents.BLOCK_ACCELERATORS,
  IpcEvents.CLICK_TITLEBAR_MAC,
  IpcEvents.OUTPUT_ENTRY,
  IpcEvents.TASK_DONE,
  IpcEvents.RELOAD_WINDOW,
];

export const ipcRendererEvents = [
  IpcEvents.OPEN_SETTINGS,
  IpcEvents.LOAD_GIST_REQUEST,
  IpcEvents.LOAD_ELECTRON_EXAMPLE_REQUEST,
  IpcEvents.FIDDLE_RUN,
  IpcEvents.FIDDLE_PACKAGE,
  IpcEvents.FIDDLE_MAKE,
  IpcEvents.MONACO_EXECUTE_COMMAND,
  IpcEvents.MONACO_TOGGLE_OPTION,
  IpcEvents.FS_NEW_FIDDLE,
  IpcEvents.FS_NEW_TEST,
  IpcEvents.FS_OPEN_FIDDLE,
  IpcEvents.FS_OPEN_TEMPLATE,
  IpcEvents.FS_SAVE_FIDDLE,
  IpcEvents.FS_SAVE_FIDDLE_AS,
  IpcEvents.FS_SAVE_FIDDLE_FORGE,
  IpcEvents.FS_SAVE_FIDDLE_GIST,
  IpcEvents.SHOW_WELCOME_TOUR,
  IpcEvents.CLEAR_CONSOLE,
  IpcEvents.LOAD_LOCAL_VERSION_FOLDER,
  IpcEvents.BISECT_COMMANDS_TOGGLE,
  IpcEvents.BEFORE_QUIT,
  IpcEvents.GET_APP_PATHS,
  IpcEvents.SELECT_ALL_IN_EDITOR,
  IpcEvents.SET_SHOW_ME_TEMPLATE,
  IpcEvents.TASK_BISECT,
  IpcEvents.TASK_TEST,
];

export const WEBCONTENTS_READY_FOR_IPC_SIGNAL =
  'WEBCONTENTS_READY_FOR_IPC_SIGNAL';
