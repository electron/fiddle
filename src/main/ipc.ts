import { EventEmitter } from 'events';

import { ipcMain } from 'electron';

import {
  IpcEvents,
  WEBCONTENTS_READY_FOR_IPC_SIGNAL,
  ipcMainEvents,
} from '../ipc-events';
import { getOrCreateMainWindow } from './windows';

/**
 * The main purpose of this class is to be the central
 * gathering place for IPC calls the main process sends
 * or listens to.
 *
 * @class IpcManager
 * @extends {EventEmitter}
 */
class IpcMainManager extends EventEmitter {
  public readyWebContents = new WeakSet<Electron.WebContents>();
  private messageQueue = new WeakMap<
    Electron.WebContents,
    Array<[IpcEvents, Array<any> | undefined]>
  >();

  constructor() {
    super();

    ipcMainEvents.forEach((name) => {
      ipcMain.removeAllListeners(name);
      ipcMain.on(name, (...args: Array<any>) => this.emit(name, ...args));
    });

    ipcMain.on(
      WEBCONTENTS_READY_FOR_IPC_SIGNAL,
      (event: Electron.IpcMainEvent) => {
        this.readyWebContents.add(event.sender);

        const queue = this.messageQueue.get(event.sender);
        this.messageQueue.delete(event.sender);
        if (!queue) return;
        for (const item of queue) {
          this.send(item[0], item[1], event.sender);
        }
      },
    );
  }

  /**
   * Send an IPC message to an instance of Electron.WebContents.
   * If none is specified, we'll automatically go with the main window.
   *
   * @param {IpcEvents} channel
   * @param {Array<any>} [args]
   * @param {Electron.WebContents} [target]
   */
  public send(
    channel: IpcEvents,
    args?: Array<any>,
    target?: Electron.WebContents,
  ) {
    const _target = target || getOrCreateMainWindow().webContents;
    const _args = args || [];
    if (!this.readyWebContents.has(_target)) {
      const existing = this.messageQueue.get(_target) || [];
      this.messageQueue.set(_target, [...existing, [channel, args]]);
      return;
    }

    _target.isDestroyed() || _target.send(channel, ..._args);
  }

  public handle(
    channel: IpcEvents,
    listener: (event: Electron.IpcMainInvokeEvent, ...args: any[]) => any,
  ) {
    // there can be only one, so remove previous one first
    ipcMain.removeHandler(channel);
    ipcMain.handle(channel, listener);
  }

  public handleOnce(
    channel: IpcEvents,
    listener: (event: Electron.IpcMainInvokeEvent, ...args: any[]) => any,
  ) {
    ipcMain.handleOnce(channel, listener);
  }
}

export const ipcMainManager = new IpcMainManager();
