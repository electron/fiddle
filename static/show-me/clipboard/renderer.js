window.addEventListener('DOMContentLoaded', () => {
  const copyButton = document.querySelector('#copy')
  const pasteButton = document.querySelector('#paste')
  const textarea = document.querySelector('textarea')
  copyButton.onclick = () => {
    window.clipboard.writeText('Hello from Electron!')
  }
  pasteButton.onclick = async () => {
    textarea.value = await window.clipboard.readText()
  }
})
