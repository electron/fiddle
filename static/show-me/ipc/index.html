<!DOCTYPE html>
<html lang="en">
<head>
  <title>IPC Example</title>
</head>
<body>
  <h1>IPC Example</h1>
  <p>Open the developer tools to see the IPC in action.</p>
</body>
</html>
