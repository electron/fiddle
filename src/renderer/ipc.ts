import { EventEmitter } from 'events';

import { ipcRenderer } from 'electron';

import { IpcEvents, ipcRendererEvents } from '../ipc-events';

/**
 * The main purpose of this class is to be the central
 * gathering place for IPC calls the main process sends
 * or listens to.
 *
 * @class IpcManager
 * @extends {EventEmitter}
 */
class IpcRendererManager extends EventEmitter {
  constructor() {
    super();

    ipcRendererEvents.forEach((name) => {
      ipcRenderer.removeAllListeners(name);
      ipcRenderer.on(name, (...args: Array<any>) => this.emit(name, ...args));
    });
  }

  /**
   * Central method to send an IPC message to the main process.
   * Currently very simple, but centralized here so that future
   * refactors will be a lot easier.
   *
   * @param {IpcEvents} channel
   * @param {...Array<any>} args
   * @memberof IpcRendererManager
   */
  public send(channel: IpcEvents, ...args: Array<any>) {
    ipcRenderer.send(channel, ...args);
  }

  public invoke(channel: IpcEvents, ...args: Array<any>) {
    return ipcRenderer.invoke(channel, ...args);
  }
}

export const ipcRendererManager = new IpcRendererManager();
