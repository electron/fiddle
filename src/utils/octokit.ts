import { Octokit } from '@octokit/rest';

import { AppState } from '../renderer/state';

let _octo: Octokit;

/**
 * Returns a loaded Octokit. If state is passed and authentication
 * is available, we'll token-authenticate.
 *
 * @export
 * @returns {Promise<typeof Octokit>}
 */
export async function getOctokit(appState?: AppState): Promise<Octokit> {
  // It's possible to load Gists without being authenticated,
  // but we get better rate limits when authenticated.
  _octo =
    _octo || appState?.gitHubToken
      ? new Octokit({
          auth: appState?.gitHubToken,
        })
      : new Octokit();

  return _octo;
}
