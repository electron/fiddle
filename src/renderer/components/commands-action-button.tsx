import * as React from 'react';

import {
  Button,
  ButtonGroup,
  IToastProps,
  Menu,
  MenuItem,
  Popover,
  Position,
  Toaster,
} from '@blueprintjs/core';
import { clipboard } from 'electron';
import { when } from 'mobx';
import { observer } from 'mobx-react';

import {
  EditorValues,
  GistActionState,
  GistActionType,
} from '../../interfaces';
import { IpcEvents } from '../../ipc-events';
import { ensureRequiredFiles } from '../../utils/editor-utils';
import { getOctokit } from '../../utils/octokit';
import { ipcRendererManager } from '../ipc';
import { AppState } from '../state';

interface GistActionButtonProps {
  appState: AppState;
}

interface IGistActionButtonState {
  readonly isUpdating: boolean;
  readonly isDeleting: boolean;
  readonly actionType: GistActionType;
}

/**
 * The "publish" button takes care of logging you in.
 *
 * @export
 * @class GistActionButton
 * @extends {React.Component<GistActionButtonProps, GistActionButtonState>}
 */
export const GistActionButton = observer(
  class GistActionButton extends React.Component<
    GistActionButtonProps,
    IGistActionButtonState
  > {
    public constructor(props: GistActionButtonProps) {
      super(props);
      this.handleClick = this.handleClick.bind(this);
      this.performGistAction = this.performGistAction.bind(this);
      this.setPrivate = this.setPrivate.bind(this);
      this.setPublic = this.setPublic.bind(this);

      this.state = {
        isUpdating: false,
        isDeleting: false,
        actionType: GistActionType.publish,
      };

      ipcRendererManager.removeAllListeners(IpcEvents.FS_SAVE_FIDDLE_GIST);
    }

    private toaster: Toaster;
    private refHandlers = {
      toaster: (ref: Toaster) => (this.toaster = ref),
    };

    public componentDidMount() {
      ipcRendererManager.on(IpcEvents.FS_SAVE_FIDDLE_GIST, this.handleClick);
    }

    public componentWillUnmount() {
      ipcRendererManager.off(IpcEvents.FS_SAVE_FIDDLE_GIST, this.handleClick);
    }

    /**
     * When the user clicks the publish button, we either show the
     * authentication dialog or publish right away.
     *
     * If we're showing the authentication dialog, we wait for it
     * to be closed again (or a GitHub token to show up) before
     * we publish
     *
     * @returns {Promise<void>}
     * @memberof GistActionButton
     */
    public async handleClick(): Promise<void> {
      const { appState } = this.props;

      if (!appState.gitHubToken) {
        appState.toggleAuthDialog();
      }

      // Wait for the dialog to be closed again
      await when(
        () => !!appState.gitHubToken || !appState.isTokenDialogShowing,
      );

      if (appState.gitHubToken) {
        return this.performGistAction();
      }
    }

    private getFiddleDescriptionFromUser(): Promise<string | undefined> {
      const placeholder = 'Electron Fiddle Gist' as const;
      return this.props.appState.showInputDialog({
        defaultInput: placeholder,
        label: 'Please provide a brief description for your Fiddle Gist',
        ok: 'Publish',
        placeholder,
      });
    }

    private async publishGist(description: string) {
      const { appState } = this.props;

      const octo = await getOctokit(appState);
      const { gitHubPublishAsPublic } = appState;
      const options = { includeDependencies: true, includeElectron: true };
      const values = await window.ElectronFiddle.app.getEditorValues(options);

      try {
        const gist = await octo.gists.create({
          public: !!gitHubPublishAsPublic,
          description,
          files: this.gistFilesList(values) as any, // Note: GitHub messed up, GistsCreateParamsFiles is an incorrect interface
        });

        appState.gistId = gist.data.id;
        appState.localPath = undefined;

        console.log(`Publish Button: Publishing complete`, { gist });
        this.renderToast({
          message: 'Successfully published gist!',
          action: {
            text: 'Copy link',
            icon: 'clipboard',
            onClick: () => clipboard.writeText(gist.data.html_url),
          },
        });

        // Only set action type to update if publish completed successfully.
        this.setActionType(GistActionType.update);
      } catch (error) {
        console.warn(`Could not publish gist`, { error });

        const messageBoxOptions: Electron.MessageBoxOptions = {
          message:
            'Publishing Fiddle to GitHub failed. Are you connected to the Internet?',
          detail: `GitHub encountered the following error: ${error.message}`,
        };

        ipcRendererManager.send(
          IpcEvents.SHOW_WARNING_DIALOG,
          messageBoxOptions,
        );
      }
    }

    /**
     * Publish a new GitHub gist.
     */
    public async handlePublish() {
      const { appState } = this.props;
      appState.activeGistAction = GistActionState.publishing;

      const description = await this.getFiddleDescriptionFromUser();

      if (description) {
        await this.publishGist(description);
        appState.editorMosaic.isEdited = false;
      }

      appState.activeGistAction = GistActionState.none;
    }

    /**
     * Update an existing GitHub gist.
     */
    public async handleUpdate() {
      const { appState } = this.props;
      const octo = await getOctokit(this.props.appState);
      const options = { includeDependencies: true, includeElectron: true };
      const values = await window.ElectronFiddle.app.getEditorValues(options);

      appState.activeGistAction = GistActionState.updating;

      try {
        const {
          data: { files: oldFiles },
        } = await octo.gists.get({ gist_id: appState.gistId! });

        const files = this.gistFilesList(values);
        for (const id of Object.keys(oldFiles)) {
          // Gist files are deleted by setting content to an empty string.
          if (!(id in files)) files[id] = { content: '' };
        }

        const gist = await octo.gists.update({
          gist_id: appState.gistId!,
          files,
        });

        appState.editorMosaic.isEdited = false;
        console.log('Updating: Updating done', { gist });
        this.renderToast({
          message: 'Successfully updated gist!',
          action: {
            text: 'Copy link',
            icon: 'clipboard',
            onClick: () => clipboard.writeText(gist.data.html_url),
          },
        });
      } catch (error) {
        console.warn(`Could not update gist`, { error });

        const messageBoxOptions: Electron.MessageBoxOptions = {
          message:
            'Updating Fiddle Gist failed. Are you connected to the Internet and is this your Gist?',
          detail: `GitHub encountered the following error: ${error.message}`,
          buttons: ['Ok'],
        };

        ipcRendererManager.send(
          IpcEvents.SHOW_WARNING_DIALOG,
          messageBoxOptions,
        );
      }

      appState.activeGistAction = GistActionState.none;
      this.setActionType(GistActionType.update);
    }

    /**
     * Delete an existing GitHub gist.
     */
    public async handleDelete() {
      const { appState } = this.props;
      const octo = await getOctokit(this.props.appState);

      appState.activeGistAction = GistActionState.deleting;

      try {
        const gist = await octo.gists.delete({
          gist_id: appState.gistId!,
        });

        appState.editorMosaic.isEdited = true;
        console.log('Deleting: Deleting done', { gist });
        this.renderToast({ message: 'Successfully deleted gist!' });
      } catch (error) {
        console.warn(`Could not delete gist`, { error });

        const messageBoxOptions: Electron.MessageBoxOptions = {
          message:
            'Deleting Fiddle Gist failed. Are you connected to the Internet, is this your Gist, and have you loaded it?',
          detail: `GitHub encountered the following error: ${error.message}`,
        };

        ipcRendererManager.send(
          IpcEvents.SHOW_WARNING_DIALOG,
          messageBoxOptions,
        );
      }

      appState.gistId = undefined;
      appState.activeGistAction = GistActionState.none;
      this.setActionType(GistActionType.publish);
    }

    /**
     * Connect with GitHub, perform a publish/update/delete action,
     * and update all related properties in the app state.
     */
    public async performGistAction(): Promise<void> {
      const { gistId } = this.props.appState;

      const actionType = gistId
        ? this.state.actionType
        : GistActionType.publish;

      switch (actionType) {
        case GistActionType.delete:
          return this.handleDelete();

        case GistActionType.publish:
          return this.handlePublish();

        case GistActionType.update:
          return this.handleUpdate();
      }
    }

    /**
     * Publish fiddles as private.
     *
     * @memberof GistActionButton
     */
    public setPrivate() {
      this.setPrivacy(false);
    }

    /**
     * Publish fiddles as public.
     *
     * @memberof GistActionButton
     */
    public setPublic() {
      this.setPrivacy(true);
    }

    public render() {
      const { gistId, activeGistAction } = this.props.appState;
      const { actionType } = this.state;

      const getTextForButton = () => {
        let text;
        if (gistId) {
          text = actionType;
        } else if (activeGistAction === GistActionState.updating) {
          text = 'Updating...';
        } else if (activeGistAction === GistActionState.publishing) {
          text = 'Publishing...';
        } else if (activeGistAction === GistActionState.deleting) {
          text = 'Deleting...';
        } else {
          text = 'Publish';
        }
        return text;
      };

      const getActionIcon = () => {
        switch (actionType) {
          case GistActionType.publish:
            return 'upload';
          case GistActionType.update:
            return 'refresh';
          case GistActionType.delete:
            return 'delete';
        }
      };

      const isPerformingAction = activeGistAction !== GistActionState.none;
      return (
        <>
          <fieldset disabled={isPerformingAction}>
            <ButtonGroup className="button-gist-action">
              {this.renderPrivacyMenu()}
              <Button
                onClick={this.handleClick}
                loading={isPerformingAction}
                icon={getActionIcon()}
                text={getTextForButton()}
              />
              {this.renderGistActionMenu()}
            </ButtonGroup>
          </fieldset>
          <Toaster
            position={Position.BOTTOM_RIGHT}
            ref={this.refHandlers.toaster}
          />
        </>
      );
    }

    private renderGistActionMenu = () => {
      const { gistId } = this.props.appState;
      const { actionType } = this.state;

      if (!gistId) {
        return null;
      }

      const menu = (
        <Menu>
          <MenuItem
            text="Publish"
            active={actionType === GistActionType.publish}
            onClick={() => this.setActionType(GistActionType.publish)}
          />
          <MenuItem
            text="Update"
            active={actionType === GistActionType.update}
            onClick={() => this.setActionType(GistActionType.update)}
          />
          <MenuItem
            text="Delete"
            active={actionType === GistActionType.delete}
            onClick={() => this.setActionType(GistActionType.delete)}
          />
        </Menu>
      );

      return (
        <Popover content={menu} position={Position.BOTTOM}>
          <Button icon="wrench" />
        </Popover>
      );
    };

    private renderPrivacyMenu = () => {
      const { gitHubPublishAsPublic, gistId } = this.props.appState;
      const { actionType } = this.state;

      if (gistId && actionType !== GistActionType.publish) {
        return null;
      }

      const privacyIcon = gitHubPublishAsPublic ? 'unlock' : 'lock';
      const privacyMenu = (
        <Menu>
          <MenuItem
            text="Private"
            icon="lock"
            active={!gitHubPublishAsPublic}
            onClick={this.setPrivate}
          />
          <MenuItem
            text="Public"
            icon="unlock"
            active={gitHubPublishAsPublic}
            onClick={this.setPublic}
          />
        </Menu>
      );

      return (
        <Popover content={privacyMenu} position={Position.BOTTOM}>
          <Button icon={privacyIcon} />
        </Popover>
      );
    };

    private setActionType = (actionType: GistActionType) => {
      this.setState({ actionType });
    };

    private setPrivacy(publishAsPublic: boolean) {
      this.props.appState.gitHubPublishAsPublic = publishAsPublic;
    }

    private renderToast = (toast: IToastProps) => {
      this.toaster?.show(toast);
    };

    private gistFilesList = (values: EditorValues) => {
      values = ensureRequiredFiles(values);
      return Object.fromEntries(
        Object.entries(values)
          .filter(([, content]) => Boolean(content))
          .map(([id, content]) => [id, { content }]),
      );
    };
  },
);
