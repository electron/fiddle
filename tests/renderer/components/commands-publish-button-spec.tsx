import * as React from 'react';

import { shallow } from 'enzyme';

import {
  EditorValues,
  GistActionState,
  GistActionType,
  MAIN_JS,
} from '../../../src/interfaces';
import { IpcEvents } from '../../../src/ipc-events';
import { GistActionButton } from '../../../src/renderer/components/commands-action-button';
import { ipcRendererManager } from '../../../src/renderer/ipc';
import { getOctokit } from '../../../src/utils/octokit';
import { AppMock, StateMock, createEditorValues } from '../../mocks/mocks';

jest.mock('../../../src/utils/octokit');

class OctokitMock {
  private static nextId = 1;

  public authenticate = jest.fn();
  public gists = {
    create: jest.fn().mockImplementation(() => ({
      data: {
        id: OctokitMock.nextId++,
      },
    })),
    delete: jest.fn(),
    update: jest.fn(),
    get: jest.fn(),
  };
}

type GistFile = { content: string };
type GistFiles = { [id: string]: GistFile };
type GistCreateOpts = {
  description: string;
  files: GistFiles;
  public: boolean;
};

describe('Action button component', () => {
  const description = 'Electron Fiddle Gist';
  const errorMessage = 'ðŸ’€';
  let app: AppMock;
  let mocktokit: OctokitMock;
  let state: StateMock;
  let expectedGistOpts: GistCreateOpts;

  function getGistFiles(values: EditorValues): GistFiles {
    return Object.fromEntries(
      Object.entries(values).map(([id, content]) => [
        id,
        { content } as GistFile,
      ]),
    );
  }

  beforeEach(() => {
    ({ app } = (window as any).ElectronFiddle);
    ({ state } = app);

    // have the octokit getter use our mock
    mocktokit = new OctokitMock();
    (getOctokit as any).mockImplementation(() => mocktokit);

    // listen for generated ipc traffic
    ipcRendererManager.send = jest.fn();

    // build ExpectedGistCreateOpts
    const editorValues = createEditorValues();
    const files = getGistFiles(editorValues);
    expectedGistOpts = { description, files, public: true } as const;
  });

  function createActionButton() {
    const wrapper = shallow(<GistActionButton appState={state as any} />);
    const instance = wrapper.instance() as any;
    return { wrapper, instance };
  }

  it('renders', () => {
    const { wrapper } = createActionButton();
    expect(wrapper).toMatchSnapshot();
  });

  it('registers for FS_SAVE_FIDDLE_GIST events', () => {
    const event = IpcEvents.FS_SAVE_FIDDLE_GIST;

    // confirm that it starts listening when mounted
    let spy = jest.spyOn(ipcRendererManager, 'on');
    const { instance, wrapper } = createActionButton();
    expect(spy).toHaveBeenCalledWith(event, instance.handleClick);
    spy.mockRestore();

    // confirm that it stops listening when unmounted
    spy = jest.spyOn(ipcRendererManager, 'off');
    wrapper.unmount();
    expect(spy).toHaveBeenCalledWith(event, instance.handleClick);
    spy.mockRestore();
  });

  it('toggles the auth dialog on click if not authed', async () => {
    state.toggleAuthDialog = jest.fn();
    const { instance } = createActionButton();
    await instance.handleClick();
    expect(state.toggleAuthDialog).toHaveBeenCalled();
  });

  it('toggles the publish method on click if authed', async () => {
    state.gitHubToken = 'github-token';

    const { instance } = createActionButton();
    instance.performGistAction = jest.fn();
    await instance.handleClick();

    expect(instance.performGistAction).toHaveBeenCalled();
  });

  describe('publish mode', () => {
    let instance: any;

    beforeEach(() => {
      // create a button that's primed to publish a new gist
      ({ instance } = createActionButton());
    });

    it('publishes a gist', async () => {
      state.showInputDialog = jest.fn().mockResolvedValueOnce(description);
      await instance.performGistAction();
      expect(mocktokit.gists.create).toHaveBeenCalledWith(expectedGistOpts);
    });

    it('asks the user for a description', async () => {
      const description = 'some non-default description';
      state.showInputDialog = jest.fn().mockResolvedValueOnce(description);
      await instance.performGistAction();
      expect(mocktokit.gists.create).toHaveBeenCalledWith({
        ...expectedGistOpts,
        description,
      });
    });

    it('publishes only if the user confirms', async () => {
      state.showInputDialog = jest.fn().mockResolvedValueOnce(undefined);
      await instance.performGistAction();
      expect(mocktokit.gists.create).not.toHaveBeenCalled();
    });

    describe('empty files', () => {
      it('are replaced with default content for required files', async () => {
        const values = { [MAIN_JS]: '' } as const;

        app.getEditorValues.mockReturnValueOnce(values);
        const { instance } = createActionButton();
        state.showInputDialog = jest.fn().mockResolvedValueOnce(description);
        await instance.performGistAction();

        const files = getGistFiles(values);
        const expected = { ...expectedGistOpts, files };
        expect(mocktokit.gists.create).toHaveBeenCalledWith(expected);
      });

      it('are omitted if they are not required files', async () => {
        const required = { [MAIN_JS]: '// fnord' };
        const optional = { 'foo.js': '' };

        app.getEditorValues.mockReturnValueOnce({ ...required, ...optional });
        const { instance } = createActionButton();
        state.showInputDialog = jest.fn().mockResolvedValueOnce(description);
        await instance.performGistAction();

        const files = getGistFiles(required);
        const expected = { ...expectedGistOpts, files };
        expect(mocktokit.gists.create).toHaveBeenCalledWith(expected);
      });
    });

    it('handles an error in Gist publishing', async () => {
      mocktokit.gists.create.mockImplementation(() => {
        throw new Error(errorMessage);
      });

      state.showInputDialog = jest.fn().mockResolvedValueOnce(description);

      const { instance } = createActionButton();
      await instance.performGistAction();
      expect(state.activeGistAction).toBe(GistActionState.none);
    });

    it('can publish private gists', async () => {
      state.showInputDialog = jest.fn().mockResolvedValueOnce(description);
      instance.setPrivate();
      await instance.performGistAction();
      const { create } = mocktokit.gists;
      expect(create).toHaveBeenCalledWith({
        ...expectedGistOpts,
        public: false,
      });
    });

    it('can publish public gists', async () => {
      state.showInputDialog = jest.fn().mockResolvedValueOnce(description);
      instance.setPublic();
      await instance.performGistAction();
      const { create } = mocktokit.gists;
      expect(create).toHaveBeenCalledWith({
        ...expectedGistOpts,
        public: true,
      });
    });
  });

  describe('update mode', () => {
    const gistId = '123';
    let wrapper: any;
    let instance: any;

    beforeEach(() => {
      // create a button that's primed to update gistId
      state.gistId = gistId;
      ({ instance, wrapper } = createActionButton());
      wrapper.setProps({ appState: state });
      instance.setState({ actionType: GistActionType.update });

      mocktokit.gists.get.mockImplementation(() => {
        return {
          data: expectedGistOpts,
        };
      });
    });

    it('attempts to update an existing Gist', async () => {
      await instance.performGistAction();

      expect(mocktokit.gists.update).toHaveBeenCalledWith({
        gist_id: gistId,
        files: expectedGistOpts.files as unknown,
      });
    });

    it('notifies the user if updating fails', async () => {
      mocktokit.gists.update.mockImplementation(() => {
        throw new Error(errorMessage);
      });

      await instance.performGistAction();

      expect(ipcRendererManager.send).toHaveBeenCalledWith(
        IpcEvents.SHOW_WARNING_DIALOG,
        expect.objectContaining({
          detail: expect.stringContaining(errorMessage),
          message: expect.stringContaining('Updating Fiddle Gist failed.'),
        }),
      );
    });
  });

  describe('delete mode', () => {
    const gistId = '123';
    let wrapper: any;
    let instance: any;

    beforeEach(() => {
      state.gistId = gistId;

      // create a button primed to delete gistId
      ({ instance, wrapper } = createActionButton());
      wrapper.setProps({ appState: state });
      instance.setState({ actionType: GistActionType.delete });
    });

    it('attempts to delete an existing Gist', async () => {
      await instance.performGistAction();
      expect(mocktokit.gists.delete).toHaveBeenCalledWith({ gist_id: gistId });
    });

    it('notifies the user if deleting fails', async () => {
      mocktokit.gists.delete.mockImplementation(() => {
        throw new Error(errorMessage);
      });

      await instance.performGistAction();

      expect(ipcRendererManager.send).toHaveBeenCalledWith(
        IpcEvents.SHOW_WARNING_DIALOG,
        expect.objectContaining({
          detail: expect.stringContaining(errorMessage),
          message: expect.stringContaining('Deleting Fiddle Gist failed.'),
        }),
      );
    });
  });

  describe('disables itself', () => {
    async function testDisabledWhen(gistActionState: GistActionState) {
      // create a button with no initial state
      state.activeGistAction = GistActionState.none;
      const { wrapper } = createActionButton();
      expect(wrapper.find('fieldset').prop('disabled')).toBe(false);

      state.activeGistAction = gistActionState;
      expect(wrapper.find('fieldset').prop('disabled')).toBe(true);

      state.activeGistAction = GistActionState.none;
      expect(wrapper.find('fieldset').prop('disabled')).toBe(false);
    }

    it('while publishing', async () => {
      await testDisabledWhen(GistActionState.publishing);
    });
    it('while updating', async () => {
      await testDisabledWhen(GistActionState.updating);
    });
    it('while deleting', async () => {
      await testDisabledWhen(GistActionState.deleting);
    });
  });

  describe('privacy menu', () => {
    it('toggles the privacy setting', () => {
      const { instance } = createActionButton();

      instance.setPublic();
      expect(state.gitHubPublishAsPublic).toBe(true);

      instance.setPrivate();
      expect(state.gitHubPublishAsPublic).toBe(false);
    });
  });
});
