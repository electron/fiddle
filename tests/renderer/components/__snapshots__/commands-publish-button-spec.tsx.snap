// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Action button component renders 1`] = `
<Fragment>
  <fieldset
    disabled={false}
  >
    <Blueprint3.ButtonGroup
      className="button-gist-action"
    >
      <Blueprint3.Popover
        boundary="scrollParent"
        captureDismiss={false}
        content={
          <Blueprint3.Menu>
            <Blueprint3.MenuItem
              active={false}
              disabled={false}
              icon="lock"
              multiline={false}
              onClick={[Function]}
              popoverProps={Object {}}
              shouldDismissPopover={true}
              text="Private"
            />
            <Blueprint3.MenuItem
              active={true}
              disabled={false}
              icon="unlock"
              multiline={false}
              onClick={[Function]}
              popoverProps={Object {}}
              shouldDismissPopover={true}
              text="Public"
            />
          </Blueprint3.Menu>
        }
        defaultIsOpen={false}
        disabled={false}
        fill={false}
        hasBackdrop={false}
        hoverCloseDelay={300}
        hoverOpenDelay={150}
        inheritDarkTheme={true}
        interactionKind="click"
        minimal={false}
        modifiers={Object {}}
        openOnTargetFocus={true}
        position="bottom"
        targetTagName="span"
        transitionDuration={300}
        usePortal={true}
        wrapperTagName="span"
      >
        <Blueprint3.Button
          icon="unlock"
        />
      </Blueprint3.Popover>
      <Blueprint3.Button
        icon="upload"
        loading={false}
        onClick={[Function]}
        text="Publish"
      />
    </Blueprint3.ButtonGroup>
  </fieldset>
  <Blueprint3.Toaster
    autoFocus={false}
    canEscapeKeyClear={true}
    position="bottom-right"
    usePortal={true}
  />
</Fragment>
`;
